tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py

```diff
@@ -121,10 +121,69 @@ def tile_batch(t, multiplier, name=None):
     return nest.map_structure(lambda t_: _tile_batch(t_, multiplier), t)
 
 
+def gather_tree_from_array(t, parent_ids, sequence_length):
+  """Calculates the full beams for `TensorArray`s.
+
+  Args:
+    t: A `TensorArray` of size `max_time` that contains `Tensor`s of shape
+      `[batch_size, beam_width, depth]`.
+    parent_ids: The parent ids of shape `[max_time, batch_size, beam_width]`.
+    sequence_length: The sequence length of shape `[batch_size, beam_width]`.
+
+  Returns:
+    A `TensorArray` of the same size and type as `t` and where beams are sorted
+    in each `Tensor` according to `parent_ids`.
+  """
+  max_time = array_ops.shape(parent_ids)[0]
+  batch_size = array_ops.shape(parent_ids)[1]
+  beam_width = array_ops.shape(parent_ids)[2]
+
+  # Generate beam ids that will be reordered by gather_tree.
+  beam_ids = array_ops.expand_dims(
+      array_ops.expand_dims(math_ops.range(beam_width), 0), 0)
+  beam_ids = array_ops.tile(beam_ids, [max_time, batch_size, 1])
+
+  mask = array_ops.sequence_mask(
+      sequence_length, maxlen=max_time, dtype=dtypes.int32)
+  mask = array_ops.transpose(mask, perm=[2, 0, 1])
+
+  # Use beam_width + 1 to mark the end of beam.
+  masked_beam_ids = (beam_ids * mask) + (1 - mask) * (beam_width + 1)
+
+  max_sequence_lengths = math_ops.to_int32(
+      math_ops.reduce_max(sequence_length, axis=1))
+  sorted_beam_ids = beam_search_ops.gather_tree(
+      step_ids=masked_beam_ids,
+      parent_ids=parent_ids,
+      max_sequence_lengths=max_sequence_lengths,
+      end_token=beam_width + 1)
+
+  # For out of range steps, simply copy the same beam.
+  sorted_beam_ids = array_ops.where(
+      math_ops.cast(mask, dtypes.bool), x=sorted_beam_ids, y=beam_ids)
+
+  # Gather from each tensor in t according to sorted_beam_ids.
+  def _collect(collector, i):
+    gathered = _tensor_gather_helper(
+        gather_indices=sorted_beam_ids[i],
+        gather_from=t.read(i),
+        batch_size=batch_size,
+        range_size=beam_width,
+        gather_shape=[batch_size * beam_width, -1])
+    return collector.write(i, gathered), i + 1
+
+  collected = tensor_array_ops.TensorArray(
+      t.dtype, size=t.size(), dynamic_size=False)
+  collected, _ = control_flow_ops.while_loop(
+      lambda _, i: i < t.size(),
```

Comment:
would it make sense here to just do a ta.stack() at each time step and run the (much faster) gather_tree op on this?  this inner loop looks REALLY slow.

add a comment in the top docstring that if requesting the alignment history, the complexity goes from `O(time_steps)` to `O(time_steps^2)`.

also add a test with `has_attention=False`, `has_alignment_history=True`.

Reply:
I'm not sure to understand. This loop is only called at the end of the decoding in `finalize()` and iterates over `Tensor`s. As  `gather_tree` works on `step_ids`, here we gather a generated sequence of ids that we then use to reorder each element of the `TensorArray`.

Am I missing something here?

Regarding the test, when `has_attention` is `False` there is no `AttentionWrapper` and no `alignment_history` by extension.

Reply:
so then my question remains the same.  why not just stack() the TensorArray and perform the reordering using something like `gather_nd`?

Reply:
(which would likely be much faster...)

Reply:
It sounds like a good idea by stacking and gather_nd, which will indeed be faster than current loop. i guess we need to unstack as well so the returned type is still tensorarray.

Reply:
How would you generate the indices for `gather_nd` efficiently? (@steven-hh-ding If you want to take a look that would be appreciated.)

The idea behind this loop is to apply the same gathering code (i.e. `_tensor_gather_helper`) on Tensors, whether they are directly in the cell state or within the TensorArray.

Reply:
if i remember the shapes here correctly:


```python
time = 3
batch = 4
beam = 5
depth = 6

# [time, batch, beam, depth]
a = tf.constant(np.arange(time * batch * beam * depth), shape=[time, batch, beam, depth], dtype=tf.int64)

# [time, batch, beam] (reverse beam as example)
beam_ids = tf.constant(np.flip(np.arange(beam), 0), shape=[1,1,beam], dtype=tf.int64)
beam_ids = tf.tile(beam_ids, [3,4,1])

# we need [time, batch, beam, 3]
# assume beam_ids > -1
t_bt_ind = tf.where(tf.ones_like(beam_ids) > -1)[:,:-1]
t_bt_ind = tf.reshape(t_bt_ind, [time,batch,beam,-1])
t_bt_be_ind = tf.concat([t_bt_ind, tf.expand_dims(beam_ids, -1)], -1)

# gather
tf.gather_nd(a, t_bt_be_ind)

```

Please double-check. Another way is to combine range and tile; but it is hardly readable. We use this method a lot in our code (tf.where(constant)), it looks okay. ebrevdo may better comment on this. Thanks!